Add useMemo util for htmlIdGenerator + update EuiDataGrid's IDs to not change on every rerender

constancecchen opened the pull request:
### Summary

@kqualters-elastic recently raised that our EuiDataGrid IDs were regenerating on every change/rerender (e.g. click/sort/paginate events, etc.). The way to fix this is to memoize the generated ID.

~I opted to create a custom hook helper with `useRef`, as the generated ID isn't really meant to change unless the component fully unmounts/mounts, and I think `useRef` conveys that slightly more accurately than state.~

Per @chandlerprall's excellent breakdown and @breehall's work, I've determined that `useMemo` definitely beats out `useRef` in terms of usage and updated this PR accordingly!

#### Before
(Note the yellow flash on the `id` props on change)

![before](https://user-images.githubusercontent.com/549407/131877093-fa0e7587-506f-4158-b1fc-68fd79f52874.gif)

#### After

![after](https://user-images.githubusercontent.com/549407/131877092-7f00647e-2a8b-417a-af88-54f9f8958c5f.gif)

### Checklist

~- [ ] Check against **all themes** for compatibility in both light and dark modes~
~- [ ] Checked in **mobile**~
~- [ ] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**~
~- [ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#adding-playground-toggles)**~

- [x] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)**
- [x] Checked **[Code Sandbox](https://codesandbox.io/)** works for the any docs examples
- [x] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**

~- [ ] Checked for **breaking changes** and labeled appropriately~

- [x] Checked for **accessibility** including keyboard-only and screenreader modes
- [x] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

constancecchen commented:
As a note, I only updated `EuiDataGrid` for now to get early comments out of the way, but I'm planning on doing a pass on all components with rerendering generated IDs. FWIW It's not all components using `htmlIdGenerator`, as some of them are either using useState/useRef already, or are class components that store the generated ID on init which also doesn't rerender. Do y'all have any preference as to whether those updates should be a separate PR, or if I should continue extending this PR?

constancecchen commented on src/services/accessibility/html_id_generator.ts:
```diff
@@ -23,3 +24,18 @@ export function htmlIdGenerator(idPrefix: string = '') {
+ * until the component using it is unmounted. This prevents IDs from
+ * being re-randomized on every component update.
+ */
+export const useHtmlIdGenerator = ({
+  idPrefix,
+  idSuffix,
+}: {
+  idPrefix?: string;
+  idSuffix?: string;
+} = {}) => {
+  return useRef<string>(htmlIdGenerator(idPrefix)(idSuffix)).current;
+};
```
Hey y'all, would super appreciate some feedback on this new util helper:

1. I'm not in love with the `useHtmlIdGenerator` naming - I'm worried it's too unspecific or it's not super accurately conveying the differences between this hook util and the original generator fn. What are people's thoughts instead on...
    1. `useHtmlIdGeneratorRef` - should I include `ref` in here to indicate it's a ref object?
    1. `useGeneratedHtmlId` (to indicate it's not returning a function but instead the final string)

2. Is the different API usage okay with anyone else (object params 2 fn args)? We can't really keep the same curried function that the current `htmlIdGenerator()()` does, because you can't curry a hook (since you can't call a hook within a callback) and it's semi-pointless to only ref the first generator function since the final string is what we want to remain static.

constancecchen replied:
Re-reading, I think my vote goes to `useGeneratedHtmlId` - any more thoughts/suggestions/objections? :)

constancecchen replied:
I also just realized https://elastic.github.io/eui/#/utilities/html-id-generator is a thing, whoops 🙈 Should I document this new use hook in there as well?

thompsongl replied:
* `useGeneratedHtmlId` 👍 
* Documentation 👍

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5133/

constancecchen commented:
👋 Hey y'all! I ended up squashing in my changes in a rebase, since there were many of them after our recent team discussion.


1. The initial setup (c5bd90b) now uses `useMemo` instead of `useRef`, and is much awesomer for it! Thanks Chandler and Bree!
    - I also renamed the fn to differentiate it more clearly from the original generator, and renamed the params from `idPrefix`->`prefix` for brevity
2. I added 5921ac7 which adds a 3rd `idFromProps` param, and an example conversion within EuiCollapsibleNav. I'm not _super_ in love with the naming of this one though, does anyone have any ideas? Here's the options I tossed around, not sure if they're any better:
    - `idFromProps`
    - `customId`
    - `overrideId`
3. I added docs in 1009e4c and would super appreciate feedback there!! It'll be my first time officially adding copy/docs examples and I'm not totally sure I nailed it 😅 

Thanks y'all! Locally, I'll keep beavering away at converting more `src/components/` to the new util!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5133/

constancecchen commented:
👋 Re-pinging for review since this is potentially blocking @breehall's PR!

cchaos commented on src-docs/src/views/html_id_generator/html_id_generator_example.js:
```diff
@@ -116,5 +122,38 @@ export const HtmlIdGeneratorExample = {
+          type: GuideSectionTypes.JS,
+          code: UseGeneratedHtmlIdSource,
+        },
+        {
+          type: GuideSectionTypes.HTML,
+          code: UseGeneratedHtmlIdHtml,
+        },
```
You actually no longer need to supply the `.HTML` configurations to the demos as we've removed this tab. We just haven't gotten around to removing it from all the docs pages yet. But it's safe to remove this for all sections on this page too if you'd like.

constancecchen replied:
This should be addressed in https://github.com/elastic/eui/pull/5133/commits/0a2a20a11522f7edcb72533c3d031588d60d89e9!

cchaos commented on src/services/accessibility/html_id_generator.ts:
```diff
@@ -23,3 +24,22 @@ export function htmlIdGenerator(idPrefix: string = '') {
+ * until the component using it is unmounted. This prevents IDs from
+ * being re-randomized on every component update.
+ */
+export const useGeneratedHtmlId = ({
```
Since we already have a `htmlIdGenerator` service, it would be great to keep the hook similarly named so that it's easier to find similar instances.
```suggestion
export const useHtmlIdGenerator = ({
```

constancecchen replied:
Ah, so I started a thread on this earlier here (https://github.com/elastic/eui/pull/5133#discussion_r701229253) on why `useHtmlIdGenerator` naming felt wrong (which I now realize I shouldn't have pre-emptively collapsed 😅)

Basically my reasoning is that the actual usage and output of the use hook vs. the original function is significantly different, so I changed the name deliberately to make that distinction clearer. Here's why they're different:

- `useGeneratedHtmlId()` at the end of the day outputs a **memoized string**. I think `generatedHtmlId` makes that clearer.
- `useHtmlIdGenerator()` outputs a **generator function** that can be called multiple different times to generate multiple IDs that will use the same prefix UUID. The hook does not and **will always generate completely different IDs**.

As an example, a few EUI components currently do something like this:

```js
import React from 'react';

import { htmlIdGenerator } from '../../services';

const generateId = htmlIdGenerator();

export class SomeComponent {
  tooltipId = generateId('tooltip');
  panelId = generateId('panel');
  // etc.
}
```

[EuiSelectable](https://github.com/elastic/eui/blob/005f9e3afd9e2e5a6cc178b3fae053a00bbf75b0/src/components/selectable/selectable.tsx#L156) is the best example I found from a quick scan that reuses the initial UUID generator to create more suffixed child IDs. Those components likely don't want to use the current use hook as is, and need to retain usage of the original generator functionality.

On the other hand, there are plenty of other components that are simple and straightforward and just need 1 single randomly generated ID, no prefix/suffix needed. For those cases, `useGeneratedHtmlId` is perfect and they can be converted to using that. But I want to be super deliberate about that choice, as the prefix/suffix/curried functionality of `htmlIdGenerator` isn't obvious at first glance, and I think distinguishing that in the naming will help prevent further confusion down the road.

cchaos replied:
There seems to then be more nuances to this hook vs the original service that may be good to detail in the docs since I just interpreted the new hook as literally just the hook version of the same service.

constancecchen replied:
Yeah, it's definitely not! To be fair though I don't think the original generator is _super_ thoroughly documented either and some of the nuances are lost 😅 For example, if you pass in the same suffix to a generator twice separately, it will output the same ID each time, which I thought was kind of a weird choice (why not just save the generated ID to a var if you want the same ID twice?)

I'll add another commit more clearly documenting the following:

- the original generator and more of its nuances
- the specific differences between the generator and the use hook
- more examples of the use hook params (primarily the custom ID prop, once we figure out naming)

constancecchen replied:
👋 I added a new section in https://github.com/elastic/eui/pull/5133/commits/e5049faefe40fbc21acfeabb2700f5b43f506dc1 with an example of creating multiple IDs from the same generator, and added a new paragraph to the hook section explaining the differences and also what it's best used for. Let me know if you think that's sufficient!

<img width="1189" alt="" src="https://user-images.githubusercontent.com/549407/132607975-c89fbfda-b1bd-45ec-805d-ea0a71b7c111.png">

cchaos replied:
🙌 Nice!

cchaos commented on src-docs/src/views/html_id_generator/html_id_generator_example.js:
```diff
@@ -116,5 +122,38 @@ export const HtmlIdGeneratorExample = {
+          <p>
+            <EuiCode>useGeneratedHtmlId</EuiCode> optionally takes{' '}
+            <EuiCode>suffix</EuiCode> and <EuiCode>prefix</EuiCode> parameters
+            with the same behavior as above, as well as an{' '}
```
```suggestion
            producing similar results as <EuiCode>htmlIdGenerator</EuiCode>, as well as an{' '}
```

constancecchen replied:
https://github.com/elastic/eui/pull/5133/commits/0a2a20a11522f7edcb72533c3d031588d60d89e9

cchaos commented on src-docs/src/views/html_id_generator/use_generated_html_id.js:
```diff
@@ -0,0 +1,41 @@
+
+  return (
+    <Fragment>
+      <EuiFlexGroup
+        justifyContent="flexStart"
+        gutterSize="m"
+        alignItems="center"
+      >
+        <EuiFlexItem grow={false}>
+          <EuiFormRow>
+            <EuiSwitch
+              label="Clicking me changes component state"
+              checked={isChecked}
+              onChange={onChange}
+            />
+          </EuiFormRow>
+        </EuiFlexItem>
+      </EuiFlexGroup>
```
NIT, but you don't need flex groups for just a single flex item unless you're trying to vertically center. And the form row does nothing either without any props.

```suggestion
       <EuiSwitch
         label="Clicking me changes component state"
         checked={isChecked}
         onChange={onChange}
       />
```

That will clean up the demo code to get to the heart of the example.

constancecchen replied:
https://github.com/elastic/eui/pull/5133/commits/0a2a20a11522f7edcb72533c3d031588d60d89e9

cchaos commented on src-docs/src/views/html_id_generator/html_id_generator_example.js:
```diff
@@ -116,5 +122,38 @@ export const HtmlIdGeneratorExample = {
+            <EuiCode>useGeneratedHtmlId</EuiCode> optionally takes{' '}
+            <EuiCode>suffix</EuiCode> and <EuiCode>prefix</EuiCode> parameters
+            with the same behavior as above, as well as an{' '}
+            <EuiCode>idFromProps</EuiCode> option for components that allow
+            end-users to set their own custom IDs.
```
I don't understand this part, and it's not shown in the example.

constancecchen replied:
It will be used by our internal components, which need to allow overriding randomized IDs from props. See [this example in the PR](https://github.com/elastic/eui/pull/5133/commits/5921ac783b7aa8a1980a13c1dd8a88cabaf2c504#diff-6af2fc183be9dd96f8dfbcedc05af823009a2f5121c262ee81d989c99685bb64).

I'm definitely 50/50 on actually documenting in our docs/examples though 🤔 While it's technically there and available, I'm not totally sure when a consumer of EUI would actually use it (unless creating their own custom component). I'm not sure what the modus operandi is for documenting args/params that are mostly EUI-internal but still available publicly - is it better to document it for transparency, or better to leave it out to reduce confusion?

cchaos replied:
So it accepts either a string or `undefined` and if it's `undefined` will return a generated ID instead of the one provided? Basically it's the conditional? I think it's definitely worth it to document this. Even just providing some prop comments directly in the hook for each prop would be really helpful.

I'm wondering if then if `idFromProps` is too specific really for what it does and could just be `id`?

Sorry but as this will probably be integral to further component creation/maintenance, I'd really want to make sure it's understandable.

constancecchen replied:
Yeah, you got it! I'm definitely not super sold on the naming either (+1 to wondering if `idFromProps` is too specific!), and was definitely looking for feedback here. I mentioned in an [earlier comment](https://github.com/elastic/eui/pull/5133#issuecomment-912160303) that I also threw around a couple other names like `customId` and `overrideId` 🤔 What do you think? Just plain `id` works for sure, I'm just worried if it won't be clear to people, like, "why would I pass in an ID when this hook is supposed to generate one for me?" But I'm hoping we can make it clear from the var name it's an optional override. So maybe `overideID` or `idOverride` would be better?

cchaos replied:
What about `conditionalId` I know that sounds long, but I think it emulates what it expects which is a variable that may or may not exist.

constancecchen replied:
Love it!!

constancecchen replied:
👋 I changed the name and added a props table in https://github.com/elastic/eui/pull/5133/commits/61f6ab8fb62a7157ba2432067236a6efcbe93e53!

<img width="602" alt="" src="https://user-images.githubusercontent.com/549407/132607508-8c22e2bc-29ed-4976-bd3f-e49b60cb5d9f.png">

Let me know what you think and if the props table is enough to clearly explain the hook options/params. I ended up still leaving it out of the actual demo JS, mostly because I think it's unlikely an end user will use it - but if you'd strongly prefer there be a demo of it just let me know!

cchaos replied:
🥇 Fantastic!

cchaos commented on src-docs/src/views/html_id_generator/html_id_generator_example.js:
```diff
@@ -26,6 +27,11 @@ const PrefixSufixSource = require('!!raw-loader!./bothPrefixSuffix');
+const UseGeneratedHtmlIdSource = require('!!raw-loader!./use_generated_html_id');
+const UseGeneratedHtmlIdHtml = renderToHtml(UseGeneratedHtmlId);
+const useGeneratedHtmlIdSnippet =
+  "useGeneratedHtmlId({ prefix: 'Some', suffix: 'id', idFromProps: id })";
```
**Optional**:

I like to think of snippets as copy/pasteable code that forces the consumer to have to update the parts that need to be customized.

Thoughts on simply something like this:

```suggestion
  "useGeneratedHtmlId({ prefix, suffix, idFromProps: id })";
```

constancecchen replied:
Ahh that's super good to know re: snippets, thanks! I'm good with this!

cchaos reviewed (commented):
I mainly just looked at the new hook and it's associated docs. Had a couple of suggestions.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5133/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5133/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5133/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5133/

thompsongl commented on src/services/accessibility/html_id_generator.testenv.ts:
```diff
@@ -6,6 +6,8 @@
  * Side Public License, v 1.
  */
 
+import { UseGeneratedHtmlIdOptions } from './html_id_generator';
```
Referencing the original file from the testenv file typically doesn't work, but this should be fine because it's just a type. Would you mind just double-checking the build output?

constancecchen replied:
I'm relatively this is still working as tests are still passing! If it weren't or were somehow importing the original file/fn, I'd be seeing snapshot failures due to mismatched IDs (which I did see originally before I added a corresponding testenv mock)

thompsongl reviewed (commented):
Love how simple this hook turned out to be!

cchaos commented on src-docs/src/views/html_id_generator/html_id_generator_reuse.js:
```diff
@@ -0,0 +1,19 @@
+
+export const HtmlIdGeneratorReuse = () => {
+  return (
+    <EuiFlexGroup gutterSize="none" alignItems="flexStart" direction="column">
+      <EuiCode>{generateId()}</EuiCode>
+      <EuiCode>{generateId()}</EuiCode>
+      <EuiCode>{generateId()}</EuiCode>
+      <EuiCode>{generateId('hello')}</EuiCode>
+      <EuiCode>{generateId('world')}</EuiCode>
+    </EuiFlexGroup>
```
More nits about flex groups. They truly are unnecessary for column layouts especially when there's no guttersize and not use EuiFlexItem within.
```suggestion
    <>
      <EuiCode>{generateId()}</EuiCode><br />
      <EuiCode>{generateId()}</EuiCode><br />
      <EuiCode>{generateId()}</EuiCode><br />
      <EuiCode>{generateId('hello')}</EuiCode><br />
      <EuiCode>{generateId('world')}</EuiCode>
    <>
```

constancecchen replied:
👍 Nice, thanks, I'll get less lazy about EuiFlex in the future! Adding this is in separate local commit, as Prettier has Opinions™️ about the formatting

constancecchen replied:
https://github.com/elastic/eui/pull/5133/commits/5a5b1670443a914d92158999be06ce7c776709c6

cchaos reviewed (approved):
👍 LGTM, only checked in Chrome though and only the Docs of the new hook.

constancecchen commented on src/services/accessibility/html_id_generator.ts:
```diff
@@ -23,3 +24,34 @@ export function htmlIdGenerator(idPrefix: string = '') {
+/**
+ * Generates an ID within a static ref object that remains static
+ * until the component using it is unmounted. This prevents IDs from
+ * being re-randomized on every component update.
```
Thanks for the eagle eyes!! Fixed in https://github.com/elastic/eui/pull/5133/commits/5a5b1670443a914d92158999be06ce7c776709c6

constancecchen commented:
I just did another quick QA pass and noticed all the CodeSandbox links on the HTML ID generator demos are broken because they export named components instead of a default function 🙈 I'll fix that here shortly!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5133/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5133/

thompsongl reviewed (approved):
LGTM!
Codesandbox links are working, except for the one with the new hook, which is to be expected before it's released.

constancecchen commented:
Ahh phew, thanks a million Greg, I was worried I'd done something wrong there! 😅 

Merging this in now!

thompsongl commented:
> Ahh phew, thanks a million Greg, I was worried I'd done something wrong there!

Yeah we currently don't have a method to use a pre-release EUI package in codesandbox, so it uses the latest release. So any new things components or services in a PR are unavailable.
